feat: implement patient controller CRUD operations #7

abdalahsalah opened the pull request:
# Pull Request: Implement CRUD Operations for Patient Controller (#7)

This PR implements the full **CRUD (Create, Read, Update, Delete)** operations for the `Patient` entity. The implementation follows the project's existing architectural patterns, ensuring separation of concerns via DTOs, Services, and Repositories.

## Changes Made

### New Files Created
* **`DTO/request/PatientCreateDTO.cs`**
* **`DTO/request/PatientUpdateDTO.cs`**
* **`DTO/response/PatientResponseDTO.cs`**
* **`Service/Interface/IPatientService.cs`**
* **`Service/Implementation/PatientService.cs`**
* **`Controllers/PatientController.cs`**

### Files Modified
* **`Program.cs`**
* **`Repository/Interface/IPatientRepository.cs`**
* **`Repository/Repositories/PatientRepository.cs`**
* **`Repository/Interface/IDoctorRepository.cs`**
* **`Repository/Repositories/DoctorRepository.cs`**

Commit:
feat: add PatientService and update repository interfaces for patient management

mahmoud-40 commented on Controllers/PatientController.cs:
```diff
@@ -0,0 +1,193 @@
+namespace BreastCancer.Controllers
+{
+    [Route("api/[controller]")]
+    [ApiController]
```
Please add a specific route for this controller. 
Like /patient

mahmoud-40 commented on Controllers/PatientController.cs:
```diff
@@ -0,0 +1,193 @@
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Error retrieving all patients");
+                return StatusCode(500, new { message = "An error occurred while retrieving patients." });
```
Avoid using 500, use bad request like the last endpoint (http 400)

mahmoud-40 commented on Controllers/PatientController.cs:
```diff
@@ -0,0 +1,193 @@
+        {
+            try
+            {
+                if (string.IsNullOrEmpty(id))
```
The validation should be on the service layer not here

mahmoud-40 commented on Controllers/PatientController.cs:
```diff
@@ -0,0 +1,193 @@
+        /// <param name="id">Patient ID</param>
+        /// <returns>Patient details</returns>
+        [HttpGet("{id}")]
+        [Authorize(Policy = "ContentAccess")]
```
What does Content Access mean? I checked the code and there’s no definition for it

mahmoud-40 commented on Controllers/PatientController.cs:
```diff
@@ -0,0 +1,193 @@
+        /// <param name="patientDto">Patient creation data</param>
+        /// <returns>Created patient</returns>
+        [HttpPost]
+        [Authorize(Policy = "MedicalAccess")]
```
Same comment as Medical Access one

mahmoud-40 commented on DTO/request/PatientCreateDTO.cs:
```diff
@@ -0,0 +1,48 @@
+
+namespace BreastCancer.DTO.request
+{
+    public class PatientCreateDTO
```
I believe this is not a registeration and the create endpoint here would be used by the admin. So we just need the minimum required data to be required (name, email, ..). The rest can be nullable. 

Also for the password I believe the password should be temporary and the user should enter it on the first login.

mahmoud-40 commented on Repository/Interface/IDoctorRepository.cs:
```diff
@@ -1,12 +1,11 @@
 {
     public interface IDoctorRepository : IGenericRepository<Doctor>
     {
+        Task<Doctor?> GetByIdAsync(string id);
```
Why not using the implemented one from the generic repo?

abdalahsalah replied:
The explicit declaration is needed because IGenericRepository<T> defines GetByIdAsync(int id) for entities with integer IDs, while Patient and Doctor use string IDs (from IdentityUser<string>).
This is method overloading (different signatures), not redundancy

mahmoud-40 commented on Service/Implementation/PatientService.cs:
```diff
@@ -0,0 +1,253 @@
+            }
+        }
+
+        private PatientResponseDTO MapToResponseDTO(Patient patient)
```
Please remove this and use automapper as agreed for any mapping.

mahmoud-40 commented on Service/Implementation/PatientService.cs:
```diff
@@ -0,0 +1,253 @@
+        }
+
+        public async Task<bool> DeletePatientAsync(string id)
+        {
```
I believe the return type should be void, no need for the bool here.

mahmoud-40 replied:
Also implemnt the hard delete functionality

Commit:
refactor:
- Replace manual mapping with AutoMapper
- Move validation to service layer
- Simplify PatientCreateDTO (only essential fields required)
- Add temporary password generation and hard delete functionality
- Standardize error responses and authorization policies

mahmoud-40 commented on Service/Implementation/PatientService.cs:
```diff
@@ -88,16 +110,16 @@ public async Task<PatientResponseDTO> CreatePatientAsync(PatientCreateDTO patien
                     ImageUrl = patientDto.ImageUrl,
-                    DateOfBirth = patientDto.DateOfBirth,
-                    Gender = patientDto.Gender,
+                    DateOfBirth = patientDto.DateOfBirth ?? DateTime.UtcNow.AddYears(-30), // Default to 30 years ago if not provided
```
Why(-30) years? I think no need for a default value here

mahmoud-40 commented on Controllers/PatientController.cs:
```diff
@@ -154,38 +158,46 @@ public async Task<IActionResult> UpdatePatient(string id, [FromBody] PatientUpda
-        [Authorize(Policy = "MedicalAccess")]
-        public async Task<IActionResult> DeletePatient(string id)
+        [Authorize(Policy = "SystemAdmin")]
+        public async Task<IActionResult> DeletePatient(string id, [FromQuery] bool hardDelete = false)
```
Make another endpoint for the hard delete

mahmoud-40 commented on Controllers/PatientController.cs:
```diff
@@ -122,17 +125,13 @@ public async Task<IActionResult> CreatePatient([FromBody] PatientCreateDTO patie
+        /// <remarks>ContentAccess policy allows access to users with roles: Doctor, Admin, Patient, or Caregiver</remarks>
         [HttpPut("{id}")]
-        [Authorize(Policy = "MedicalAccess")]
+        [Authorize(Policy = "ContentAccess")]
```
No one should update the patient except the patient abd admin role

mahmoud-40 commented on keycloak/import/breast-cancer-realm.json:
```diff
@@ -7,19 +7,19 @@
     "realm": [
       {
-        "name": "patient",
+        "name": "Patient",
```
We already canceled keycloak so no need for these edits

Commit:
refactor: update authorization policies and patient model properties

mahmoud-40 commented on Controllers/PatientController.cs:
```diff
@@ -167,21 +167,14 @@ public async Task<IActionResult> UpdatePatient(string id, [FromBody] PatientUpda
         [HttpDelete("{id}")]
-        [Authorize(Policy = "SystemAdmin")]
-        public async Task<IActionResult> DeletePatient(string id, [FromQuery] bool hardDelete = false)
+        [Authorize(Policy = "AdminOrPatient")]
```
يسطا وربنا انت صعب

احنا لو كل endpoint هنعملها هنعملها بوليسي هنا ف كده مش هنخلص

استخدم احسن
[Authorize(Roles = "Admin,Patient")]

وبلاش نستخدم حوار البوليسي دي خالص

mahmoud-40 commented on Controllers/PatientController.cs:
```diff
@@ -200,6 +193,38 @@ public async Task<IActionResult> DeletePatient(string id, [FromQuery] bool hardD
+        /// <returns>No content on success</returns>
+        /// <remarks>SystemAdmin policy allows access only to users with role: Admin</remarks>
+        [HttpDelete("{id}/HardDelete")]
+        [Authorize(Policy = "SystemAdmin")]
```
نفس الكلام هنا
ايه الفرق بين ال system admin وال admin
اتسخدم admin عطول

mahmoud-40 reviewed (approved):
Please resolve the comments before merging

Commit:
refactor: update authorization attributes in PatientController to use role-based access

mahmoud-40 reviewed (approved):
اسطوري والله 
يارب تتجوز
